python/ray/serve/long_pull.py

```diff
@@ -0,0 +1,148 @@
+import asyncio
+import random
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import Any, Awaitable, Callable, DefaultDict, Dict, Generic, Set, Tuple, List, TypeVar, Optional
+
+import ray
+from ray.serve.utils import logger
+
+
+@dataclass
+class UpdatedObject:
+    object_snapshot: Any
+    snapshot_id: int
+
+
+UpdateStateCallable = Callable[[Dict[str, Any], List[str]], None]
+UpdateStateAsyncCallable = Callable[[Dict[str, Any], List[str]], Awaitable[
+    None]]
+
+
+class BaseClient:
+    def __init__(self, host_actor, keys: List[str]) -> None:
+        self.host_actor = host_actor
+        self.snapshot_ids: Dict[str, int] = {key: -1 for key in keys}
```

Comment:
Does -1 signify that we've never seen the option before or something?

Reply:
Oh, I see -- it's signifying the version number which starts at zero so we'll always pull the first version. Maybe leave a comment.